src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -114,7 +125,19 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                     if (step.getId().equals(neededDatum.getDataId())) {
                         if (entityScreen.referencesContainStep(step.getValue())) {
                             currentStep = step.getValue();
+                            processedSteps.add(step);
                             needsFullInit = ++processedStepsCount == steps.size();
+                        } else {
+                            // This block constructs the message to display then throws the exception
```

Comment:
similar to the comment above, it would be good to extract this to a method

Reply:
[6754f61](https://github.com/dimagi/formplayer/pull/1560/commits/6754f61adb48a77d3581fcce4be805c8385918e6)